File: common/src/main/java/juuxel/adorn/criterion/BoughtFromTradingStationCriterion.java
package juuxel.adorn.criterion;

import com.mojang.serialization.Codec;
import com.mojang.serialization.codecs.RecordCodecBuilder;
import net.minecraft.advancements.predicates.ContextAwarePredicate;
import net.minecraft.advancements.predicates.ItemPredicate;
import net.minecraft.advancements.predicates.entity.EntityPredicate;
import net.minecraft.advancements.triggers.SimpleCriterionTrigger;
import net.minecraft.server.level.ServerPlayer;
import net.minecraft.world.item.ItemStack;

import java.util.Optional;

public final class BoughtFromTradingStationCriterion extends SimpleCriterionTrigger<BoughtFromTradingStationCriterion.Conditions> {
    @Override
    public Codec<Conditions> codec() {
        return Conditions.CODEC;
    }

    public void trigger(ServerPlayer player, ItemStack soldItem) {
        trigger(player, conditions -> conditions.matches(soldItem));
    }

    public record Conditions(Optional<ContextAwarePredicate> player, Optional<ItemPredicate> soldItem) implements SimpleCriterionTrigger.SimpleInstance {
        public static final Codec<Conditions> CODEC = RecordCodecBuilder.create(instance -> instance.group(
            EntityPredicate.ADVANCEMENT_CODEC.optionalFieldOf("player")
                .forGetter(Conditions::player),
            ItemPredicate.CODEC.optionalFieldOf("item")
                .forGetter(Conditions::soldItem)
        ).apply(instance, Conditions::new));

        public boolean matches(ItemStack stack) {
            return soldItem.map(predicate -> predicate.test(stack)).orElse(true);
        }
    }
}

File: common/src/main/kotlin/juuxel/adorn/criterion/BoughtFromTradingStationCriterion.kt
package juuxel.adorn.criterion

import com.mojang.serialization.Codec
import com.mojang.serialization.codecs.RecordCodecBuilder
import net.minecraft.advancement.criterion.AbstractCriterion
import net.minecraft.item.ItemStack
import net.minecraft.predicate.entity.EntityPredicate
import net.minecraft.predicate.entity.LootContextPredicate
import net.minecraft.predicate.item.ItemPredicate
import net.minecraft.server.network.ServerPlayerEntity
import net.minecraft.util.dynamic.Codecs
import java.util.Optional

class BoughtFromTradingStationCriterion : AbstractCriterion<BoughtFromTradingStationCriterion.Conditions>() {
    override fun getConditionsCodec(): Codec<Conditions> = Conditions.CODEC

    fun trigger(player: ServerPlayerEntity, soldItem: ItemStack) {
        trigger(player) { it.matches(soldItem) }
    }

    data class Conditions(val playerPredicate: Optional<LootContextPredicate>, val soldItem: Optional<ItemPredicate>) : AbstractCriterion.Conditions {
        override fun player(): Optional<LootContextPredicate> = playerPredicate

        fun matches(stack: ItemStack): Boolean =
            soldItem.map { it.test(stack) }.orElse(true)

        companion object {
            val CODEC: Codec<Conditions> = RecordCodecBuilder.create { instance ->
                instance.group(
                    Codecs.createStrictOptionalFieldCodec(EntityPredicate.LOOT_CONTEXT_PREDICATE_CODEC, "player")
                        .forGetter(Conditions::playerPredicate),
                    Codecs.createStrictOptionalFieldCodec(ItemPredicate.CODEC, "item")
                        .forGetter(Conditions::soldItem)
                ).apply(instance, ::Conditions)
            }
        }
    }
}
